File: Leetcode July Challenge/C++/Find Median from Data Stream.cpp
#include<bits/stdc++.h>
class MedianFinder {
public:
    priority_queue<int> maxpq; //maxHeap
    priority_queue<int, vector<int>, greater<int>> minpq; //minHeap
    
    void addNum(int num) {
            
        if(maxpq.empty() || maxpq.top() > num){
            maxpq.push(num);
        }else{
            minpq.push(num);
        }
        
        if(maxpq.size() > minpq.size() + 1){
            minpq.push(maxpq.top());
            maxpq.pop();
        }
        else if(minpq.size() > maxpq.size() + 1){
            maxpq.push(minpq.top());
            minpq.pop();
        }
        
    }
    
    double findMedian() {
       if(maxpq.size() == minpq.size()){
           return (maxpq.top() + minpq.top()) / 2.0;
       }
        else{
            if(maxpq.size() > minpq.size()){
                return maxpq.top();
            }else{
                return minpq.top();
            }
        }
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: Leetcode July Challenge/Python/Find Median from Data Stream.py
from heapq import *

class MedianFinder:

    def __init__(self):
        self.min_heap: List[int] = []
        self.max_heap: List[int] = []
        

    def addNum(self, num: int) -> None:
        heappush(self.max_heap, -heappushpop(self.min_heap, num))
        
        if len(self.max_heap) > len(self.min_heap):
            heappush(self.min_heap, -heappop(self.max_heap))
        

    def findMedian(self) -> float:
        has_even_count = len(self.max_heap) == len(self.min_heap)

        if has_even_count:
            return (-self.max_heap[0] + self.min_heap[0]) / 2.0
        return float(self.min_heap[0])
